src/disk_io_thread.cpp

```diff
@@ -1123,6 +1131,24 @@ namespace libtorrent
 
 		m_stats_counters.inc_stats_counter(counters::num_running_disk_jobs, -1);
 
+		mutex::scoped_lock l(m_cache_mutex);
+		int check_ret = check_cache_level(l, completed_jobs);
+		l.unlock();
+
+		if (check_ret == retry_job)
+		{
+			// we only need to post a job if there are no other jobs in the queue
+			// otherwise check_cache_level will be called again after the next job
```

Comment:
this comment made me think a bit more about this. if I understand correctly, let's say thread 1 is flushing dirty blocks to disk. thread 2 runs check_cache_level and discovers we need to flush and evict more blocks. It can evict read blocks that nobody is referencing, but if it needs to flush buffers to disk, it will make sure no other thread is currently writing.
The reason for this is that the other thread that's writing, could be writing the same piece we're looking at. But that means we'll never use more than one thread to write ti disk. Doesnt't that seem strange?

if I understand the semantics correctly, the idea behavior to only have a single thread at a time do writing, we could just have a counter, every time you want to flush, you do an atomic increment. If the old value was zero, you go ahead and flush, if the previous value was non-zero, you just keep going. no need to post any job.

Once you're done flushing, you do an atomic decrement. If the new value is not zero, you simply loop back to the beginning again and call check_cache_level() again (without incrementing). This way, whichever thread is flushing, it will keep flushing until it has invoked the flushing function the correct number of times.

The risk with posting the retry job is that you essentially end up busy-spinning (except you accept new jobs while spinning), waiting for the other thread to complete.